### PR TITLE
Make import stripping smarter to resolve #5019

### LINKED ISSUE #5019: Could not find module `ember-data/-private` imported from `ember-data/setup-container`
I'm currently unable to get my project to build (again). I tried to `yarn add` a [dependency](https://github.com/Igor-Vladyka/leaflet.browser.print) - I don't know what it did, but it did something. I don't see any changes to package.json or bower.json in my working copy (it added a yarn.lock but I reverted everything for sanity). Here are those files:

```
{
  "name": "cityxcity",
  "version": "0.0.0",
  "description": "Small description for cityxcity goes here",
  "private": true,
  "directories": {
    "doc": "doc",
    "test": "tests"
  },
  "scripts": {
    "build": "ember build",
    "start": "ember server",
    "test": "ember test",
    "deploy": "ember github-pages:commit --message \"Deploy gh-pages from commit $(git rev-parse HEAD)\" --environment=production; git push origin gh-pages:gh-pages -f; git checkout develop"
  },
  "repository": "",
  "engines": {
    "node": ">= 0.12.0"
  },
  "author": "",
  "license": "MIT",
  "devDependencies": {
    "broccoli-asset-rev": "^2.4.5",
    "ember-ajax": "^2.4.1",
    "ember-array-contains-helper": "1.3.1",
    "ember-async-image": "0.1.1",
    "ember-browserify": "^1.1.13",
    "ember-c3": "0.3.1",
    "ember-c3-shim": "0.0.6",
    "ember-cli": "2.9.1",
    "ember-cli-app-version": "^2.0.0",
    "ember-cli-autoprefixer": "0.6.0",
    "ember-cli-babel": "^5.1.7",
    "ember-cli-data-export": "0.1.28",
    "ember-cli-dependency-checker": "^1.3.0",
    "ember-cli-deploy": "0.5.1",
    "ember-cli-eslint": "3.0.0",
    "ember-cli-form-data": "1.0.0",
    "ember-cli-github-pages": "0.1.2",
    "ember-cli-htmlbars": "^1.0.10",
    "ember-cli-htmlbars-inline-precompile": "^0.3.3",
    "ember-cli-inject-live-reload": "^1.4.1",
    "ember-cli-mirage": "0.2.4",
    "ember-cli-mocha": "0.11.0",
    "ember-cli-moment-shim": "3.0.1",
    "ember-cli-nouislider": "0.11.0",
    "ember-cli-release": "^0.2.9",
    "ember-cli-sri": "^2.1.0",
    "ember-cli-string-helpers": "1.3.0",
    "ember-cli-test-loader": "^1.1.0",
    "ember-cli-uglify": "^1.2.0",
    "ember-composable-helpers": "1.1.2",
    "ember-d3": "0.3.2",
    "ember-data": "^2.11.0",
    "ember-export-application-global": "^1.0.5",
    "ember-form-for": "2.0.0-alpha.4",
    "ember-leaflet": "3.0.9",
    "ember-leaflet-cartodb": "git://github.com/allthesignals/ember-leaflet-cartodb.git",
    "ember-leaflet-marker-cluster": "0.0.5",
    "ember-load-initializers": "^0.5.1",
    "ember-mocha": "0.9.3",
    "ember-modal-dialog": "0.9.0",
    "ember-moment": "7.3.0",
    "ember-pagefront": "0.11.2",
    "ember-pikaday": "2.2.1",
    "ember-power-select": "1.5.0-beta.2",
    "ember-resolver": "^2.0.3",
    "ember-route-action-helper": "2.0.2",
    "ember-searchable-select": "0.10.0",
    "ember-simple-auth": "1.2.0-beta.1",
    "ember-string-helpers": "1.0.2",
    "ember-template-lint": "^0.6.3",
    "ember-tether": "0.4.1",
    "ember-truth-helpers": "1.3.0",
    "ember-uploader": "1.2.3",
    "ember-welcome-page": "^1.0.3",
    "ember-wormhole": "0.5.1",
    "eslint-plugin-ember-suave": "^1.0.0",
    "liquid-fire": "0.27.0",
    "liquid-tether": "2.0.3",
    "loader.js": "^4.0.10",
    "semantic-ui-ember": "2.0.0",
    "squel": "^5.5.0"
  },
  "dependencies": {
    "ember-cli-sass": "^6.0.0",
    "geojson2svg": "^1.0.6",
    "string-humanize": "^1.0.0"
  }
}
```

```
{
  "name": "cityxcity",
  "dependencies": {
    "ember": "~2.9.0",
    "ember-cli-shims": "~0.1.1",
    "ember-mocha-adapter": "~0.3.1",
    "semantic-ui": "^2.2.6",
    "leaflet": "1.0.2",
    "nouislider": "^8.0.1",
    "leaflet.markercluster": "^1.0.0",
    "d3": "3.5.17",
    "c3": "^0.4.11",
    "file-saver": "^1.3.3",
    "js-xlsx": "~0.8.0"
  },
  "devDependencies": {
    "squel": "^5.5.0"
  },
  "resolutions": {
    "leaflet": ">= 1.0.0"
  }
}
```

I can't clone and stand my project up anymore either. Either something happened between the first time I installed a package (as it was in my local node_modules) and when I re-installed that package, or something is happening locally that I can't understand. 

Reverting back to to previous working commits on my project doesn't seem to help. 

Here are some probably-relevant NPM warns:
```
npm WARN ember-data@2.14.0 requires a peer of ember-inflector@^2.0.0 but none was installed.
npm WARN broccoli-typescript-compiler@0.6.2 requires a peer of typescript@^1.6.2 || ^1.7.0 || ^1.8.0 || ^1.9.0-dev || ^2.0.0 || ^2.0.0-dev || ^2.1.0-dev || next but none was installed.
npm WARN broccoli-tslinter@2.0.1 requires a peer of typescript@^2.0.0 but none was installed.
npm WARN broccoli-tslinter@2.0.1 requires a peer of tslint@^4.0.2 but none was installed.
```

I will try to upgrade my version of ember-data, but I am conservative about changing versions because of the number of dependencies I have. I wasn't able to find this particular issue elsewehre, so sorry if it already exists - I think there may be a bug.

### PR DESCRIPTION
**Pull request by runspired**
Previously we would only strip import declarations when all imports from a given module were previously filtered out by another plugin.  This PR allows us to strip filtered imports from declarations without stripping the entire declaration if other imports are still needed.

### CODE REVIEW

**Comment by stefanpenner** [comment]
LGTM, lets add a quick node-test covering this.

**Comment by xomaczar** [comment]
@stefanpenner @runspired can this be merge to resolve #5016 also in 2.15.0

**Comment by stefanpenner** [comment]
@xomaczar unfortunately, not until it has tests. cc @runspired

**Comment by xomaczar** [comment]
I don't mind pushing it forward, if u can clarify what kind of tests you are looking for. I assume something similar to the current e-d generators

**Comment by stefanpenner** [comment]
@xomaczar a unit test, which uses babel directly, and configures it to use the above plugin, on various inputs confirming functionality works as expected.

**Comment by xomaczar** [comment]
@stefanpenner thanks for the info @runspired thx for impl

**Comment by stefanpenner** [comment]
@runspired thanks for the test! It does look like the new test is failing on CI, mind taking a look?

```
1) Unit: babel-plugin-remove-filtered-imports Strips :
     TypeError: unknown: Cannot read property 'name' of undefined
      at PluginPass.ImportDeclaration (lib/transforms/babel-plugin-remove-imports.js:36:58)
      at newFn (node_modules/babel-traverse/lib/visitors.js:276:21)
      at NodePath._call (node_modules/babel-traverse/lib/path/context.js:76:18)
      at NodePath.call (node_modules/babel-traverse/lib/path/context.js:48:17)
      at NodePath.visit (node_modules/babel-traverse/lib/path/context.js:105:12)
      at TraversalContext.visitQueue (node_modules/babel-traverse/lib/context.js:150:16)
      at TraversalContext.visitMultiple (node_modules/babel-traverse/lib/context.js:103:17)
      at TraversalContext.visit (node_modules/babel-traverse/lib/context.js:190:19)
      at Function.traverse.node (node_modules/babel-traverse/lib/index.js:114:17)
      at NodePath.visit (node_modules/babel-traverse/lib/path/context.js:115:19)
```

**Comment by runspired** [comment]
@stefanpenner figured I'd push my working state in case others wanted to contribute. I figured out why the test fails though, babel output doesn't create the same AST for string transpile as it does for file transpile. Talked with @rwjblue and he pointed me to broccoli-test-helpers. Getting that working now.

**Comment by runspired** [comment]
@stefanpenner r?

**Comment by runspired** [comment]
hrm, something is up with the build

**Comment by runspired** [comment]
The appveyor log suggests it was actually successful ...

File: lib/transforms/babel-plugin-remove-imports.js [hunk 1]
```diff
@@ -24,17 +24,48 @@ function removeImports() {
...
+
+            for (let i = 0; i < imports.length; i++) {
+              if (imports[i].type === 'ImportNamespaceSpecifier') {
+                continue;
+              }
+
+              let specifier = imports[i].imported;
+
+              if (!specifier) {
+                debugger;
```
Comment (stefanpenner) [inline lib/transforms/babel-plugin-remove-imports.js]: i suspect this doesn't want to be here

**Comment by xomaczar** [comment]
Merge?

**Comment by stefanpenner** [comment]
@locks are you able to backport and release as well?

**Comment by locks** [comment]
@stefanpenner I'll only be able to do it on Sunday, if there's urgency take it away 💪 otherwise I'll tackle it.

**Comment by quaertym** [comment]
When is the expected release?

**Comment by locks** [comment]
Sorry, haven't had free weekends to handle this. I put up a backport PR at https://github.com/emberjs/data/pull/5188.

**Pull request merged**